src/mpid/ch4/shm/ipc/src/ipc_win.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "ipc_noinline.h"
+#include "ipc_mem.h"
+
+/* Return global node rank of each process in the shared communicator.
+ * I.e., rank in MPIR_Process.comm_world->node_comm. The caller routine
+ * must allocate/free each buffer. */
+static int get_node_ranks(MPIR_Comm * shm_comm_ptr, int *shm_ranks, int *node_ranks)
+{
+    int i;
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Group *shm_group_ptr;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_GET_NODE_RANKS);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_GET_NODE_RANKS);
+
+    for (i = 0; i < shm_comm_ptr->local_size; i++)
+        shm_ranks[i] = i;
+
+    mpi_errno = MPIR_Comm_group_impl(shm_comm_ptr, &shm_group_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* Get node group if it is not yet initialized */
+    if (!MPIDI_IPC_global.node_group_ptr) {
+        mpi_errno = MPIR_Comm_group_impl(MPIR_Process.comm_world->node_comm,
+                                         &MPIDI_IPC_global.node_group_ptr);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+    mpi_errno = MPIR_Group_translate_ranks_impl(shm_group_ptr, shm_comm_ptr->local_size,
+                                                shm_ranks, MPIDI_IPC_global.node_group_ptr,
+                                                node_ranks);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    mpi_errno = MPIR_Group_free_impl(shm_group_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_GET_NODE_RANKS);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+
+typedef struct IPC_win_shared_info {
+    uint32_t disp_unit;
+    size_t size;
+    MPIDI_SHM_IPC_type_t ipc_type;
+    MPIDI_IPC_mem_handle_t mem_handle;
+} IPC_win_shared_info_t;
+
+int MPIDI_IPC_mpi_win_create_hook(MPIR_Win * win)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_IPC_win_t *ipc_win = NULL;
+    MPIR_Comm *shm_comm_ptr = win->comm_ptr->node_comm;
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+    int i;
+    size_t total_shm_size = 0;
+    MPIDIG_win_shared_info_t *shared_table = NULL;
+    IPC_win_shared_info_t *ipc_shared_table = NULL;     /* temporary exchange buffer */
+    int *ranks_in_shm_grp = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IPC_MPI_WIN_CREATE_HOOK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IPC_MPI_WIN_CREATE_HOOK);
+    MPIR_CHKPMEM_DECL(2);
+    MPIR_CHKLMEM_DECL(2);
+
+    if (!shm_comm_ptr)
+        goto fn_exit;
+
+    /* Determine submodule availability and IPC type based on buffer type.
+     * If no submodule is available, skip initialization; if any submodule
+     * is on, we need always exchange in case remote buffer can be shared
+     * by IPC. We pass MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE as data_sz to
+     * MPIDI_IPC_get_mem_attr becuase for RMA operations we always enable
```

Comment:
```suggestion
     * MPIDI_IPC_get_mem_attr because for RMA operations we always enable
```

Reply:
This entire comment seems out of place.

Reply:
The comment is outdated. We should update it.